#! /usr/bin/env python
# -*- coding: utf-8 -*-
import xlrd
import time
import sys
import test_pb2
import codecs
import os

def file_20(num):
    test_20=test_pb2.TestMsg20()
    test_20.field1='1940795147'
    test_20.field2='1940795147'
    test_20.field3='1940795147'
    test_20.field4='1940795147'
    test_20.field5='1940795147'
    test_20.field6='1940795147'
    test_20.field7='1940795147'
    test_20.field8='1940795147'
    test_20.field9='1940795147'
    test_20.field10='1940795147'
    test_20.field11='1940795147'
    test_20.field12='1940795147'
    test_20.field13='1940795147'
    test_20.field14='1940795147'
    test_20.field15='1940795147'
    test_20.field16='1940795147'
    test_20.field17='1940795147'
    test_20.field18='1940795147'
    test_20.field19='1940795147'
    test_20.field20='1940795147'
    begin_time = time.time()
    data = test_20.SerializeToString()
    mid_time = time.time()
    target = test_pb2.TestMsg20()
    target.ParseFromString(data)
    end_time = time.time()
    if(num==1):
        print('Size after Protobuf encoding：'+str(sys.getsizeof(data))+'B')
    return mid_time-begin_time,end_time-mid_time

def  file_100(num):
    test_100=test_pb2.TestMsg100()
    test_100.field1=31415926
    test_100.field2=61415923
    test_100.field3=31415269
    test_100.field4=53141926
    test_100.field5=13145926
    test_100.field6=43115926
    test_100.field7=31419265
    test_100.field8=23141596
    test_100.field9=43161592
    test_100.field10=31412112
    test_100.field11=31415926
    test_100.field12=61415923
    test_100.field13=31415269
    test_100.field14=53141926
    test_100.field15=13145926
    test_100.field16=43115926
    test_100.field17=31419265
    test_100.field18=23141596
    test_100.field19=43161592
    test_100.field20=31412112
    test_100.field21=31415926
    test_100.field22=61415923
    test_100.field23=31415269
    test_100.field24=53141926
    test_100.field25=13145926
    test_100.field26=43115926
    test_100.field27=31419265
    test_100.field28=23141596
    test_100.field29=43161592
    test_100.field30=31412112
    test_100.field31=31415926
    test_100.field32=61415923
    test_100.field33=31415269
    test_100.field34=53141926
    test_100.field35=13145926
    test_100.field36=43115926
    test_100.field37=31419265
    test_100.field38=23141596
    test_100.field39=43161592
    test_100.field40=31412112
    test_100.field41=31.415926
    test_100.field42=61.415923
    test_100.field43=31.415269
    test_100.field44=53.141926
    test_100.field45=13.145926
    test_100.field46=43.115926
    test_100.field47=31.419265
    test_100.field48=23.141596
    test_100.field49=43.161592
    test_100.field50=31.412112
    test_100.field51=31.415926
    test_100.field52=61.415923
    test_100.field53=31.415269
    test_100.field54=53.141926
    test_100.field55=13.145926
    test_100.field56=43.115926
    test_100.field57=31.419265
    test_100.field58=23.141596
    test_100.field59=43.161592
    test_100.field60=31.412112
    test_100.field61=31.415926
    test_100.field62=61.415923
    test_100.field63=31.415269
    test_100.field64=53.141926
    test_100.field65=13.145926
    test_100.field66=43.115926
    test_100.field67=31.419265
    test_100.field68=23.141596
    test_100.field69=43.161592
    test_100.field70=31.412112
    test_100.field71='1a2b3c4d5f'
    test_100.field72='5f4d3c2b1a'
    test_100.field73='12345678.0'
    test_100.field74='87654321.0'
    test_100.field75='24681357.0'
    test_100.field76='abcdefgh'
    test_100.field77='hgfedcba'
    test_100.field78='acegbdfh'
    test_100.field79='1234abcd'
    test_100.field80='5678efgh'
    test_100.field81='1a2b3c4d5f'
    test_100.field82='5f4d3c2b1a'
    test_100.field83='12345678.0'
    test_100.field84='87654321.0'
    test_100.field85='24681357.0'
    test_100.field86='abcdefgh'
    test_100.field87='hgfedcba'
    test_100.field88='acegbdfh'
    test_100.field89='1234abcd'
    test_100.field90='5678efgh'
    test_100.field91='1a2b3c4d5f'
    test_100.field92='5f4d3c2b1a'
    test_100.field93='12345678.0'
    test_100.field94='87654321.0'
    test_100.field95='24681357.0'
    test_100.field96='abcdefgh'
    test_100.field97='hgfedcba'
    test_100.field98='acegbdfh'
    test_100.field99='1234abcd'
    test_100.field100='5678efgh'
    begin_time = time.time()
    data = test_100.SerializeToString()
    mid_time = time.time()
    target = test_pb2.TestMsg100()
    target.ParseFromString(data)
    end_time = time.time()
    if(num==1):
        print('Size after Protobuf encoding：'+str(sys.getsizeof(data))+'B')
    return mid_time-begin_time,end_time-mid_time

def  file_500(num):
    test_500=test_pb2.TestMsg500()
    test_500.field1=31415926
    test_500.field2=61415923
    test_500.field3=31415269
    test_500.field4=53141926
    test_500.field5=13145926
    test_500.field6=43115926
    test_500.field7=31419265
    test_500.field8=23141596
    test_500.field9=43161592
    test_500.field10=31412112
    test_500.field11=31415926
    test_500.field12=61415923
    test_500.field13=31415269
    test_500.field14=53141926
    test_500.field15=13145926
    test_500.field16=43115926
    test_500.field17=31419265
    test_500.field18=23141596
    test_500.field19=43161592
    test_500.field20=31412112
    test_500.field21=31415926
    test_500.field22=61415923
    test_500.field23=31415269
    test_500.field24=53141926
    test_500.field25=13145926
    test_500.field26=43115926
    test_500.field27=31419265
    test_500.field28=23141596
    test_500.field29=43161592
    test_500.field30=31412112
    test_500.field31=31415926
    test_500.field32=61415923
    test_500.field33=31415269
    test_500.field34=53141926
    test_500.field35=13145926
    test_500.field36=43115926
    test_500.field37=31419265
    test_500.field38=23141596
    test_500.field39=43161592
    test_500.field40=31412112
    test_500.field41=31415926
    test_500.field42=61415923
    test_500.field43=31415269
    test_500.field44=53141926
    test_500.field45=13145926
    test_500.field46=43115926
    test_500.field47=31419265
    test_500.field48=23141596
    test_500.field49=43161592
    test_500.field50=31412112
    test_500.field51=31415926
    test_500.field52=61415923
    test_500.field53=31415269
    test_500.field54=53141926
    test_500.field55=13145926
    test_500.field56=43115926
    test_500.field57=31419265
    test_500.field58=23141596
    test_500.field59=43161592
    test_500.field60=31412112
    test_500.field61=31415926
    test_500.field62=61415923
    test_500.field63=31415269
    test_500.field64=53141926
    test_500.field65=13145926
    test_500.field66=43115926
    test_500.field67=31419265
    test_500.field68=23141596
    test_500.field69=43161592
    test_500.field70=31412112
    test_500.field71=31415926
    test_500.field72=61415923
    test_500.field73=31415269
    test_500.field74=53141926
    test_500.field75=13145926
    test_500.field76=43115926
    test_500.field77=31419265
    test_500.field78=23141596
    test_500.field79=43161592
    test_500.field80=31412112
    test_500.field81=31415926
    test_500.field82=61415923
    test_500.field83=31415269
    test_500.field84=53141926
    test_500.field85=13145926
    test_500.field86=43115926
    test_500.field87=31419265
    test_500.field88=23141596
    test_500.field89=43161592
    test_500.field90=31412112
    test_500.field91=31415926
    test_500.field92=61415923
    test_500.field93=31415269
    test_500.field94=53141926
    test_500.field95=13145926
    test_500.field96=43115926
    test_500.field97=31419265
    test_500.field98=23141596
    test_500.field99=43161592
    test_500.field100=31412112
    test_500.field101=31415926
    test_500.field102=61415923
    test_500.field103=31415269
    test_500.field104=53141926
    test_500.field105=13145926
    test_500.field106=43115926
    test_500.field107=31419265
    test_500.field108=23141596
    test_500.field109=43161592
    test_500.field110=31412112
    test_500.field111=31415926
    test_500.field112=61415923
    test_500.field113=31415269
    test_500.field114=53141926
    test_500.field115=13145926
    test_500.field116=43115926
    test_500.field117=31419265
    test_500.field118=23141596
    test_500.field119=43161592
    test_500.field120=31412112
    test_500.field121=31415926
    test_500.field122=61415923
    test_500.field123=31415269
    test_500.field124=53141926
    test_500.field125=13145926
    test_500.field126=43115926
    test_500.field127=31419265
    test_500.field128=23141596
    test_500.field129=43161592
    test_500.field130=31412112
    test_500.field131=31415926
    test_500.field132=61415923
    test_500.field133=31415269
    test_500.field134=53141926
    test_500.field135=13145926
    test_500.field136=43115926
    test_500.field137=31419265
    test_500.field138=23141596
    test_500.field139=43161592
    test_500.field140=31412112
    test_500.field141=31415926
    test_500.field142=61415923
    test_500.field143=31415269
    test_500.field144=53141926
    test_500.field145=13145926
    test_500.field146=43115926
    test_500.field147=31419265
    test_500.field148=23141596
    test_500.field149=43161592
    test_500.field150=31412112
    test_500.field151=31415926
    test_500.field152=61415923
    test_500.field153=31415269
    test_500.field154=53141926
    test_500.field155=13145926
    test_500.field156=43115926
    test_500.field157=31419265
    test_500.field158=23141596
    test_500.field159=43161592
    test_500.field160=31412112
    test_500.field161=31415926
    test_500.field162=61415923
    test_500.field163=31415269
    test_500.field164=53141926
    test_500.field165=13145926
    test_500.field166=43115926
    test_500.field167=31419265
    test_500.field168=23141596
    test_500.field169=43161592
    test_500.field170=31412112
    test_500.field171=31415926
    test_500.field172=61415923
    test_500.field173=31415269
    test_500.field174=53141926
    test_500.field175=13145926
    test_500.field176=43115926
    test_500.field177=31419265
    test_500.field178=23141596
    test_500.field179=43161592
    test_500.field180=31412112
    test_500.field181=31415926
    test_500.field182=61415923
    test_500.field183=31415269
    test_500.field184=53141926
    test_500.field185=13145926
    test_500.field186=43115926
    test_500.field187=31419265
    test_500.field188=23141596
    test_500.field189=43161592
    test_500.field190=31412112
    test_500.field191=31415926
    test_500.field192=61415923
    test_500.field193=31415269
    test_500.field194=53141926
    test_500.field195=13145926
    test_500.field196=43115926
    test_500.field197=31419265
    test_500.field198=23141596
    test_500.field199=43161592
    test_500.field200=31412112
    test_500.field201=31.415926
    test_500.field202=61.415923
    test_500.field203=31.415269
    test_500.field204=53.141926
    test_500.field205=13.145926
    test_500.field206=43.115926
    test_500.field207=31.419265
    test_500.field208=23.141596
    test_500.field209=43.161592
    test_500.field210=31.412112
    test_500.field211=31.415926
    test_500.field212=61.415923
    test_500.field213=31.415269
    test_500.field214=53.141926
    test_500.field215=13.145926
    test_500.field216=43.115926
    test_500.field217=31.419265
    test_500.field218=23.141596
    test_500.field219=43.161592
    test_500.field220=31.412112
    test_500.field221=31.415926
    test_500.field222=61.415923
    test_500.field223=31.415269
    test_500.field224=53.141926
    test_500.field225=13.145926
    test_500.field226=43.115926
    test_500.field227=31.419265
    test_500.field228=23.141596
    test_500.field229=43.161592
    test_500.field230=31.412112
    test_500.field231=31.415926
    test_500.field232=61.415923
    test_500.field233=31.415269
    test_500.field234=53.141926
    test_500.field235=13.145926
    test_500.field236=43.115926
    test_500.field237=31.419265
    test_500.field238=23.141596
    test_500.field239=43.161592
    test_500.field240=31.412112
    test_500.field241=31.415926
    test_500.field242=61.415923
    test_500.field243=31.415269
    test_500.field244=53.141926
    test_500.field245=13.145926
    test_500.field246=43.115926
    test_500.field247=31.419265
    test_500.field248=23.141596
    test_500.field249=43.161592
    test_500.field250=31.412112
    test_500.field251=31.415926
    test_500.field252=61.415923
    test_500.field253=31.415269
    test_500.field254=53.141926
    test_500.field255=13.145926
    test_500.field256=43.115926
    test_500.field257=31.419265
    test_500.field258=23.141596
    test_500.field259=43.161592
    test_500.field260=31.412112
    test_500.field261=31.415926
    test_500.field262=61.415923
    test_500.field263=31.415269
    test_500.field264=53.141926
    test_500.field265=13.145926
    test_500.field266=43.115926
    test_500.field267=31.419265
    test_500.field268=23.141596
    test_500.field269=43.161592
    test_500.field270=31.412112
    test_500.field271=31.415926
    test_500.field272=61.415923
    test_500.field273=31.415269
    test_500.field274=53.141926
    test_500.field275=13.145926
    test_500.field276=43.115926
    test_500.field277=31.419265
    test_500.field278=23.141596
    test_500.field279=43.161592
    test_500.field280=31.412112
    test_500.field281=31.415926
    test_500.field282=61.415923
    test_500.field283=31.415269
    test_500.field284=53.141926
    test_500.field285=13.145926
    test_500.field286=43.115926
    test_500.field287=31.419265
    test_500.field288=23.141596
    test_500.field289=43.161592
    test_500.field290=31.412112
    test_500.field291=31.415926
    test_500.field292=61.415923
    test_500.field293=31.415269
    test_500.field294=53.141926
    test_500.field295=13.145926
    test_500.field296=43.115926
    test_500.field297=31.419265
    test_500.field298=23.141596
    test_500.field299=43.161592
    test_500.field300=31.412112
    test_500.field301=31.415926
    test_500.field302=61.415923
    test_500.field303=31.415269
    test_500.field304=53.141926
    test_500.field305=13.145926
    test_500.field306=43.115926
    test_500.field307=31.419265
    test_500.field308=23.141596
    test_500.field309=43.161592
    test_500.field310=31.412112
    test_500.field311=31.415926
    test_500.field312=61.415923
    test_500.field313=31.415269
    test_500.field314=53.141926
    test_500.field315=13.145926
    test_500.field316=43.115926
    test_500.field317=31.419265
    test_500.field318=23.141596
    test_500.field319=43.161592
    test_500.field320=31.412112
    test_500.field321=31.415926
    test_500.field322=61.415923
    test_500.field323=31.415269
    test_500.field324=53.141926
    test_500.field325=13.145926
    test_500.field326=43.115926
    test_500.field327=31.419265
    test_500.field328=23.141596
    test_500.field329=43.161592
    test_500.field330=31.412112
    test_500.field331=31.415926
    test_500.field332=61.415923
    test_500.field333=31.415269
    test_500.field334=53.141926
    test_500.field335=13.145926
    test_500.field336=43.115926
    test_500.field337=31.419265
    test_500.field338=23.141596
    test_500.field339=43.161592
    test_500.field340=31.412112
    test_500.field341=31.415926
    test_500.field342=61.415923
    test_500.field343=31.415269
    test_500.field344=53.141926
    test_500.field345=13.145926
    test_500.field346=43.115926
    test_500.field347=31.419265
    test_500.field348=23.141596
    test_500.field349=43.161592
    test_500.field350=31.412112
    test_500.field351=31.415926
    test_500.field352=61.415923
    test_500.field353=31.415269
    test_500.field354=53.141926
    test_500.field355=13.145926
    test_500.field356=43.115926
    test_500.field357=31.419265
    test_500.field358=23.141596
    test_500.field359=43.161592
    test_500.field360=31.412112
    test_500.field361=31.415926
    test_500.field362=61.415923
    test_500.field363=31.415269
    test_500.field364=53.141926
    test_500.field365=13.145926
    test_500.field366=43.115926
    test_500.field367=31.419265
    test_500.field368=23.141596
    test_500.field369=43.161592
    test_500.field370=31.412112
    test_500.field371=31.415926
    test_500.field372=61.415923
    test_500.field373=31.415269
    test_500.field374=53.141926
    test_500.field375=13.145926
    test_500.field376=43.115926
    test_500.field377=31.419265
    test_500.field378=23.141596
    test_500.field379=43.161592
    test_500.field380=31.412112
    test_500.field381=31.415926
    test_500.field382=61.415923
    test_500.field383=31.415269
    test_500.field384=53.141926
    test_500.field385=13.145926
    test_500.field386=43.115926
    test_500.field387=31.419265
    test_500.field388=23.141596
    test_500.field389=43.161592
    test_500.field390=31.412112
    test_500.field391=31.415926
    test_500.field392=61.415923
    test_500.field393=31.415269
    test_500.field394=53.141926
    test_500.field395=13.145926
    test_500.field396=43.115926
    test_500.field397=31.419265
    test_500.field398=23.141596
    test_500.field399=43.161592
    test_500.field400=31.412112
    test_500.field401='1a2b3c4d5f'
    test_500.field402='5f4d3c2b1a'
    test_500.field403='12345678.0'
    test_500.field404='87654321.0'
    test_500.field405='24681357.0'
    test_500.field406='abcdefgh'
    test_500.field407='hgfedcba'
    test_500.field408='acegbdfh'
    test_500.field409='1234abcd'
    test_500.field410='5678efgh'
    test_500.field411='1a2b3c4d5f'
    test_500.field412='5f4d3c2b1a'
    test_500.field413='12345678.0'
    test_500.field414='87654321.0'
    test_500.field415='24681357.0'
    test_500.field416='abcdefgh'
    test_500.field417='hgfedcba'
    test_500.field418='acegbdfh'
    test_500.field419='1234abcd'
    test_500.field420='5678efgh'
    test_500.field421='1a2b3c4d5f'
    test_500.field422='5f4d3c2b1a'
    test_500.field423='12345678.0'
    test_500.field424='87654321.0'
    test_500.field425='24681357.0'
    test_500.field426='abcdefgh'
    test_500.field427='hgfedcba'
    test_500.field428='acegbdfh'
    test_500.field429='1234abcd'
    test_500.field430='5678efgh'
    test_500.field431='1a2b3c4d5f'
    test_500.field432='5f4d3c2b1a'
    test_500.field433='12345678.0'
    test_500.field434='87654321.0'
    test_500.field435='24681357.0'
    test_500.field436='abcdefgh'
    test_500.field437='hgfedcba'
    test_500.field438='acegbdfh'
    test_500.field439='1234abcd'
    test_500.field440='5678efgh'
    test_500.field441='1a2b3c4d5f'
    test_500.field442='5f4d3c2b1a'
    test_500.field443='12345678.0'
    test_500.field444='87654321.0'
    test_500.field445='24681357.0'
    test_500.field446='abcdefgh'
    test_500.field447='hgfedcba'
    test_500.field448='acegbdfh'
    test_500.field449='1234abcd'
    test_500.field450='5678efgh'
    test_500.field451='1a2b3c4d5f'
    test_500.field452='5f4d3c2b1a'
    test_500.field453='12345678.0'
    test_500.field454='87654321.0'
    test_500.field455='24681357.0'
    test_500.field456='abcdefgh'
    test_500.field457='hgfedcba'
    test_500.field458='acegbdfh'
    test_500.field459='1234abcd'
    test_500.field460='5678efgh'
    test_500.field461='1a2b3c4d5f'
    test_500.field462='5f4d3c2b1a'
    test_500.field463='12345678.0'
    test_500.field464='87654321.0'
    test_500.field465='24681357.0'
    test_500.field466='abcdefgh'
    test_500.field467='hgfedcba'
    test_500.field468='acegbdfh'
    test_500.field469='1234abcd'
    test_500.field470='5678efgh'
    test_500.field471='1a2b3c4d5f'
    test_500.field472='5f4d3c2b1a'
    test_500.field473='12345678.0'
    test_500.field474='87654321.0'
    test_500.field475='24681357.0'
    test_500.field476='abcdefgh'
    test_500.field477='hgfedcba'
    test_500.field478='acegbdfh'
    test_500.field479='1234abcd'
    test_500.field480='5678efgh'
    test_500.field481='1a2b3c4d5f'
    test_500.field482='5f4d3c2b1a'
    test_500.field483='12345678.0'
    test_500.field484='87654321.0'
    test_500.field485='24681357.0'
    test_500.field486='abcdefgh'
    test_500.field487='hgfedcba'
    test_500.field488='acegbdfh'
    test_500.field489='1234abcd'
    test_500.field490='5678efgh'
    test_500.field491='1a2b3c4d5f'
    test_500.field492='5f4d3c2b1a'
    test_500.field493='12345678.0'
    test_500.field494='87654321.0'
    test_500.field495='24681357.0'
    test_500.field496='abcdefgh'
    test_500.field497='hgfedcba'
    test_500.field498='acegbdfh'
    test_500.field499='1234abcd'
    test_500.field500='5678efgh'
    begin_time = time.time() 
    data = test_500.SerializeToString()
    mid_time = time.time()
    target = test_pb2.TestMsg500()
    target.ParseFromString(data)
    end_time = time.time()
    if(num==1):
        print('Size after Protobuf encoding：'+str(sys.getsizeof(data))+'B')
    return mid_time-begin_time,end_time-mid_time

time1=0
time2=0
for i in range(10000):
    t1,t2=file_20(i)
    time1+=t1
    time2+=t2
print('20 fields Protobuf encoding time：'+str(time1/10000*1000)+'ms')
print('20 fields Protobuf decoding time：'+str(time2/10000*1000)+'ms')
print('---------------------------------')

time3=0
time4=0
for i in range(10000):
    t3,t4=file_100(i)
    time3+=t3
    time4+=t4
print('100 fields Protobuf encoding time：'+str(time3/10000*1000)+'ms')
print('100 fields Protobuf decoding time：'+str(time4/10000*1000)+'ms')
print('---------------------------------')

time5=0
time6=0
for i in range(10000):
    t5,t6=file_500(i)
    time5+=t5
    time6+=t6
print('500 fields Protobuf encoding time：'+str(time5/10000*1000)+'ms')
print('500 fields Protobuf decoding time：'+str(time6/10000*1000)+'ms')
